### ISSUE TITLE
Why is my transformer performing poorly?

### ISSUE THREAD

**Issue by erip** [pytorch/fairseq#1701]
#### What is your question?

Why is my transformer consuming so much memory _and_ performing poorly?

I am running an experiment using OpenNMT and fairseq. I'm running my OpenNMT with a config not unlike [this](https://opennmt.net/OpenNMT-py/FAQ.html#how-do-i-use-the-transformer-model) and am achieving ~40 BLEU on a GPU with 16GB of RAM. When I run a similar experiment with fairseq (with more RAM!), I run out of memory (even with memory efficient fp16 🙀). If I make my `--batch-size` small enough, eventually I don't encounter OOMs, but training is too slow to be practical (~45-60 min/epoch) and I only achieve ~20 BLEU. Am I misunderstanding something fundamental? 

#### Code

```
fairseq-train data-bin \
    --max-epoch 100 \
    --arch transformer \
    --optimizer adam --adam-betas '(0.9, 0.98)' \
    --lr 0.0005 --lr-scheduler inverse_sqrt --min-lr '1e-09' \
    --warmup-updates 4000 --warmup-init-lr '1e-07' \
    --label-smoothing 0.1 --criterion label_smoothed_cross_entropy \
    --memory-efficient-fp16 \
    --dropout 0.3 --weight-decay 0.0001 \
    --max-tokens 4096 
```

#### What have you tried?

Playing with `--batch-size`, changing embedding dimensions, plaing with `--max-tokens`, ...

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0) 1.3.1
 - OS (e.g., Linux): Windows
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): `pip install -e .`
 - Python version: 3.6.x
 - CUDA/cuDNN version: 10.1/7
 - GPU models and configuration: Titan RTX 24GB
 - Any other relevant information:

**Comment by lematt1991** [comment]
Just to clarify, are the OOMs that you're referring to GPU or CPU memory?

**Comment by lematt1991** [comment]
If I train a model on [IWSLT'14](https://github.com/pytorch/fairseq/tree/master/examples/translation#training-a-new-model) with the same command you provided, nvidia-smi shows 3097MiB of GPU memory, and top shows 22.106g of VIRT usage.  What are you seeing with openNMT?

**Comment by erip** [comment]
It's definitely GPU memory. Will poke more on OpenNMT

**Comment by lematt1991** [comment]
Hmm, it's very strange that the exact same command would utilize only ~4gb of GPU memory on my end, but would cause on OOM on your 24gb mem GPU.  Can you create a gist with the entire stdout/stderr log when you run the command you provided.

**Comment by erip** [comment]
I wonder if perhaps there's a difference in the way data is batched between windows and Linux...

**Comment by erip** [comment]
I can try to get some stdout cap tomorrow

**Comment by myleott** [comment]
That definitely seems odd, please keep us updated :)

**Comment by erip** [comment]
I haven't been able to get the literal stdout cap, but I was able to play with it a little bit. I found one issue, which was that my Titan RTX was in WDDM mode instead of TCC, so ~500MB of RAM were being used for Windows display purposes. Once I converted that to TCC and used my Quadro for display, I have been able to run (though successfully remains to be seen) `fairseq-train` with the above params (only with `--memory-efficient-fp16`). `nvidia-smi` shows **23.6GiB of RAM** usage at any given time. The number of params being trained is ~680M.
